internal/pkg/template/override/rule.go

```diff
@@ -3,36 +3,198 @@
 		}
-		node, err := r.parse()
-		if err != nil {
-			return nil, err
+	}
+	// If the node pair doesn't exist, create the label node first and then a map node.
+	// Finally we return the created map node.
+	newLabelNode := &yaml.Node{
+		Kind:  yaml.ScalarNode,
+		Tag:   nodeTagStr,
+		Value: m.key,
+	}
+	content.Content = append(content.Content, newLabelNode)
+	newValNode := &yaml.Node{
+		Kind: yaml.MappingNode,
+		Tag:  nodeTagMap,
+	}
+	content.Content = append(content.Content, newValNode)
+	return newValNode, nil
+}
+
+func (m *mapUpsertNode) upsertValue(content *yaml.Node) {
+	// If the node pair exists, substitute with the value node.
+	for i := 0; i < len(content.Content); i += 2 {
+		if m.key == content.Content[i].Value {
+			content.Content[i+1] = m.valueToInsert
 		}
-		ruleNodes = append(ruleNodes, node)
 	}
-	return ruleNodes, nil
+	// Otherwise, we create the label node then append the value node.
+	newLabelNode := &yaml.Node{
+		Kind:  yaml.ScalarNode,
+		Tag:   nodeTagStr,
+		Value: m.key,
+	}
+	content.Content = append(content.Content, newLabelNode)
+	content.Content = append(content.Content, m.valueToInsert)
 }
 
-func (r Rule) validate() error {
-	return nil
+// seqIdxUpsertNode represents a sequence node that needs to be upserted at index.
+type seqIdxUpsertNode struct {
+	index        int
+	appendToLast bool
+	upsertNode
 }
 
-func (r Rule) parse() (*ruleNode, error) {
-	return nil, nil
+// Upsert upserts a node into given yaml content.
+func (s *seqIdxUpsertNode) Upsert(content *yaml.Node) (*yaml.Node, error) {
+	// If it contains the value to insert, upsert the value to the yaml.
+	if s.valueToInsert != nil {
+		return nil, s.upsertValue(content)
+	}
+	// If the node pair exists, we check if we need to append the node to the end.
+	// If so, create a map node and return it since we want to go deeper to upsert the value.
+	// Here we assume it is not possible for the yaml we want to override to have nested sequence:
+	// Mapping01:
+	//   - - foo
+	//     - bar
+	//   - - boo
+	// The example above will be translated to "Mapping01[0][1]" to refer to "bar".
+	// If not check if the given index is within the sequence range.
+	for i := 0; i < len(content.Content); i += 2 {
+		if content.Content[i].Value == s.key {
+			seqNode := content.Content[i+1]
+			if s.appendToLast {
+				newMapNode := &yaml.Node{
+					Kind: yaml.MappingNode,
+					Tag:  nodeTagMap,
+				}
+				seqNode.Content = append(seqNode.Content, newMapNode)
+				return newMapNode, nil
+			}
+			if s.index < len(seqNode.Content) {
+				return seqNode.Content[s.index], nil
+			} else {
+				return nil, fmt.Errorf("cannot specify %s[%d] because the current length is %d. Use [%s] to append to the sequence instead",
+					s.key, s.index, len(seqNode.Content), seqAppendToLastSymbol)
+			}
+		}
+	}
+	// If the node pair doesn't exist, create the sequence node pair and a map node.
+	// Then return the created map node, since we want to go deeper to upsert the value.
```

Comment:
Under which scenario would we be creating this? `a.b[-]`? where `b` doesn't exist?
Do we need to do any validation here like checking if `appendToLast` is set?

Reply:
yeah. that's a good idea